src/main/java/de/tum/in/www1/artemis/domain/FileUploadSubmission.java

```diff
@@ -3,24 +3,48 @@
 import java.io.Serializable;
 import java.util.Objects;
 
-import javax.persistence.Column;
-import javax.persistence.DiscriminatorValue;
-import javax.persistence.Entity;
+import javax.persistence.*;
+
+import de.tum.in.www1.artemis.config.Constants;
+import de.tum.in.www1.artemis.service.FileService;
 
 /**
- * A FileUploadSubmission.
+ * NOTE: The file management is necessary to differentiate between temporary and used files and to delete used files when the corresponding submission is deleted. The workflow is
+ * as follows: 1. user uploads a file -> temporary file is created and at this point we don't know if submission is already created, when submission is created file is moved to permanent location =>
+ * This happens in @PreUpdate 2. When submission is deleted, the file in the permanent location is deleted => This happens in @PostRemove
  */
 @Entity
 @DiscriminatorValue(value = "F")
 public class FileUploadSubmission extends Submission implements Serializable {
 
     private static final long serialVersionUID = 1L;
 
+    @Transient
+    private FileService fileService = new FileService();
+
     @Column(name = "file_path")
     private String filePath;
 
     // jhipster-needle-entity-add-field - JHipster will add fields here, do not remove
 
+    @PreUpdate
+    public void onUpdate() {
+        // move file and delete old file if necessary
+        filePath = fileService.manageFilesForUpdatedFilePath(null, filePath, FileUploadSubmission.buildFilePath(getParticipation().getExercise().getId(), getId()), getId(), true);
+    }
+
+    /**
+     * Deletes solution file for this submission
+     */
+    @PostRemove
+    public void onDelete() {
+        if (filePath != null) {
+            // delete old file if necessary
+            String[] path = filePath.split("/");
+            fileService.manageFilesForUpdatedFilePath(filePath, null, FileUploadSubmission.buildFilePath(Long.parseLong(path[4]), getId()), getId(), true);
```

Comment:
this looks like code that can easily fail, how do you make sure that the path actually has 4 entries? Please write safer code!

Reply:
I think this should also really not depend on the exerciseId in the filePath itself, but rather just give the whole path to a delete method in the FileService. Then again, the actual path right now in the database is not the path in the filesystem, but the API path e.g. "/api/files/file-upload-exercises/30/submissions/100/some_file.jar". I think this should be something like "uploads/file-upload-exercises/30/submissions/100/some_file.jar"

Reply:
For now, I've added some checks and variables. But I really think we should remove the tight coupling between URLs and getting IDs for exercises, submissions, etc. in general.